pkg/util/limiter/query_limiter.go

```diff
@@ -84,6 +84,27 @@ func (ql *QueryLimiter) AddSeries(seriesLabels []cortexpb.LabelAdapter) error {
 	return nil
 }
 
+// AddSeriesBatch adds the batch of input series and returns an error if the limit is reached.
+func (ql *QueryLimiter) AddSeriesBatch(series [][]cortexpb.LabelAdapter) error {
+	// If the max series is unlimited just return without managing map
+	if ql.maxSeriesPerQuery == 0 {
+		return nil
+	}
+
+	ql.uniqueSeriesMx.Lock()
+	defer ql.uniqueSeriesMx.Unlock()
+	for _, s := range series {
+		fingerprint := client.FastFingerprint(s)
+		ql.uniqueSeries[fingerprint] = struct{}{}
+	}
```

Comment:
I am okay with this but I am also wondering if we can calculate the hashes first before we lock.
This requires to allocate a slice of int64 so not sure if it is worth. Probably we need some benchmark here

Reply:
I did see some improvement when hashing the series outside the lock. I'll implement your suggestion.

Reply:
@harry671003 Can you run the benchmark again? And let's also add `-benchmem` to show memory allocation?

Reply:
Without the slice of fps:
```
go test -bench=^BenchmarkQueryLimiter_AddSeriesBatch$ -benchmem -benchtime=200000x github.com/cortexproject/cortex/pkg/util/limiter
goos: darwin
goarch: arm64
pkg: github.com/cortexproject/cortex/pkg/util/limiter
BenchmarkQueryLimiter_AddSeriesBatch-10           200000             13591 ns/op           30929 B/op        400 allocs/op
PASS
ok      github.com/cortexproject/cortex/pkg/util/limiter        3.149s

```


With the slice of fps:
```
go test -bench=^BenchmarkQueryLimiter_AddSeriesBatch$ -benchmem -benchtime=200000x github.com/cortexproject/cortex/pkg/util/limiter
goos: darwin
goarch: arm64
pkg: github.com/cortexproject/cortex/pkg/util/limiter
BenchmarkQueryLimiter_AddSeriesBatch-10           200000             12390 ns/op           31729 B/op        401 allocs/op
PASS
ok      github.com/cortexproject/cortex/pkg/util/limiter        3.476s
```